applications/StructuralMechanicsApplication/custom_elements/truss_elements/linear_truss_element.cpp

```diff
@@ -210,81 +273,162 @@ void LinearTrussElement2D<TNNodes>::GetFirstDerivativesShapeFunctionsValues(
+        n.clear();
+        n[2] = 1.0;
     }
 
-    if (std::abs(angle) > std::numeric_limits<double>::epsilon()) {
-        BoundedMatrix<double, DofsPerNode, DofsPerNode> T;
-        BoundedMatrix<double, SystemSize, SystemSize> global_size_T;
+    // Gram-Schmidt ortogonalization
+    n = n - inner_prod(t, n) / inner_prod(t, t) * t;
+    n /= norm_2(n);
+
+    noalias(m) = MathUtils<double>::CrossProduct(t, n);
+
+    T(0, 0) = t[0];
+    T(0, 1) = t[1];
+    T(0, 2) = t[2];
+
+    T(1, 0) = n[0];
+    T(1, 1) = n[1];
+    T(1, 2) = n[2];
+
+    T(2, 0) = m[0];
+    T(2, 1) = m[1];
+    T(2, 2) = m[2];
+
+    return T;
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<SizeType TNNodes, SizeType TDimension>
+void LinearTrussElement<TNNodes, TDimension>::GetNodalValuesVector(SystemSizeBoundedArrayType& rNodalValues) const
+{
+    if (rNodalValues.size() != SystemSize)
+        rNodalValues.resize(SystemSize, false);
+    const auto &r_geom = GetGeometry();
+    BoundedVector<double, SystemSize> global_values;
+    BoundedMatrix<double, DofsPerNode, DofsPerNode> T;
+    BoundedMatrix<double, SystemSize, SystemSize> global_size_T;
+
+    if constexpr (Dimension == 2) {
+
+        const double angle = GetAngle();
+        // We fill the vector with global values
+        for (SizeType i = 0; i < NNodes; ++i) {
+            const auto& r_displ = r_geom[i].FastGetSolutionStepValue(DISPLACEMENT);
+            global_values[i * DofsPerNode]     = r_displ[0];
+            global_values[i * DofsPerNode + 1] = r_displ[1];
+        }
+
         StructuralMechanicsElementUtilities::BuildRotationMatrixForTruss(T, angle);
         if constexpr (NNodes == 2) {
             StructuralMechanicsElementUtilities::BuildElementSizeRotationMatrixFor2D2NTruss(T, global_size_T);
         } else {
             StructuralMechanicsElementUtilities::BuildElementSizeRotationMatrixFor2D3NTruss(T, global_size_T);
         }
-        // We rotate to local axes
         noalias(rNodalValues) = prod(trans(global_size_T), global_values);
     } else {
-        noalias(rNodalValues) = global_values;
+        // We fill the vector with global values
+        for (SizeType i = 0; i < NNodes; ++i) {
+            const auto& r_displ = r_geom[i].FastGetSolutionStepValue(DISPLACEMENT);
+            global_values[i * DofsPerNode]     = r_displ[0];
+            global_values[i * DofsPerNode + 1] = r_displ[1];
+            global_values[i * DofsPerNode + 2] = r_displ[2];
+        }
+
+        noalias(T) = GetFrenetSerretMatrix(); // global to local
+        if constexpr (NNodes == 2) {
+            StructuralMechanicsElementUtilities::BuildElementSizeRotationMatrixFor3D2NTruss(T, global_size_T);
+        } else {
+            StructuralMechanicsElementUtilities::BuildElementSizeRotationMatrixFor3D3NTruss(T, global_size_T);
+        }
```

Comment:
The global_size_T matrix consists of the number of dimensions blocks of matrix T. Using e.g. 
 a for loop over TNNodes with as body `MathUtils::AddMatrix(global_size_T, T, inode*TDimension, inode*TDimension)` this may be constructed without the need for separate implementations for the 2 or 3 node element.